src/caffe/proto/caffe.proto

```diff
@@ -2,13 +2,21 @@ syntax = "proto2";
 
 package caffe;
 
+// Specifies the shape (dimensions) of a Blob.
+message BlobShape {
+  repeated int64 dim = 1 [packed = true];
```

Comment:
My feeling is that it's probably worth future-proofing given the relatively tiny amount of extra storage cost (unless you're really using blobs with `INT_MAX` axes, as are now allowed...).  I could make it `uint32` for now and then later change it to `int64` if/when we really find practical uses for blobs that large, but that might cause backward compatibility problems with (de)serialization that I don't even want to think about...  (Why not `uint64`? I think it's worth keeping -1 and other <0 values reserved for various purposes -- e.g. setting `axis [default = -1]` to signify "all axes", or that an axis must be manually specified and should fail otherwise, etc. And it *does* seem pretty unlikely that in the reasonably foreseeable future we'll need 2^64 byte blobs, so 2^63-1 bytes seems like an okay max...)

I'm open to discussion on any of this though.